File: src/game.py
"""
Main game class that ties everything together.
"""

from src.empire import Empire, Colony
from src.galaxy import Galaxy
from src.combat import Ship, Fleet, Combat, Weapon, ShipClass
from src.resources import ResourceType


class Game:
    """Main game controller"""
    
    def __init__(self, empire_name="Earth Empire"):
        self.empire = Empire(empire_name)
        self.galaxy = Galaxy(num_systems=20)
        self.running = False
        
        # Setup starting conditions
        self._setup_game()
    
    def _setup_game(self):
        """Setup initial game state"""
        # Colonize home planet (Earth)
        home_planet = self.galaxy.home_system.planets[2]  # 3rd planet
        home_planet.planet_type = "Terrestrial"
        earth_colony = Colony("Earth", home_planet)
        earth_colony.population = 100
        home_planet.colonize(earth_colony)
        self.empire.add_colony(earth_colony)
        
        # Create starting fleet
        starting_fleet = self._create_starting_fleet()
        self.empire.add_fleet(starting_fleet)
    
    def _create_starting_fleet(self):
        """Create the starting fleet"""
        fleet = Fleet("Home Defense Fleet", self.empire.name)
        
        # Add basic ships
        laser = Weapon("Laser Cannon", damage=10, accuracy=0.7, range_val=5)
        
        scout = Ship("Scout-1", ShipClass.SCOUT, hull=50, shields=20, weapons=[laser])
        corvette = Ship("Corvette-1", ShipClass.CORVETTE, hull=100, shields=50, 
                       weapons=[laser, laser])
        
        fleet.add_ship(scout)
        fleet.add_ship(corvette)
        fleet.location = self.galaxy.home_system
        
        return fleet
    
    def advance_turn(self):
        """Advance the game one turn"""
        message = self.empire.advance_turn()
        
        # Colony growth
        for colony in self.empire.colonies:
            colony.grow()
        
        return message
    
    def explore_system(self, system_name):
        """Explore a star system"""
        for system in self.galaxy.systems:
            if system.name.lower() == system_name.lower():
                system.explore()
                return f"Explored {system.name}! Found {len(system.planets)} planets."
        return "System not found"
    
    def start_research(self, tech_id):
        """Start researching a technology"""
        if self.empire.set_research(tech_id):
            tech = self.empire.research.get_tech(tech_id)
            return f"Now researching: {tech.name}"
        return "Cannot research that technology"
    
    def get_available_research(self):
        """Get list of available technologies to research"""
        return self.empire.research.get_available_techs()
    
    def build_ship(self, ship_class, fleet_name):
        """Build a new ship and add to fleet"""
        # Find fleet
        fleet = None
        for f in self.empire.fleets:
            if f.name.lower() == fleet_name.lower():
                fleet = f
                break
        
        if not fleet:
            return "Fleet not found"
        
        # Simple ship creation (costs not implemented yet)
        ship_name = f"{ship_class}-{len(fleet.ships) + 1}"
        
        if ship_class == ShipClass.SCOUT:
            laser = Weapon("Laser", 10, 0.7, 5)
            ship = Ship(ship_name, ship_class, 50, 20, [laser])
        elif ship_class == ShipClass.CORVETTE:
            laser = Weapon("Laser", 10, 0.7, 5)
            ship = Ship(ship_name, ship_class, 100, 50, [laser, laser])
        elif ship_class == ShipClass.FRIGATE:
            laser = Weapon("Laser", 15, 0.75, 6)
            ship = Ship(ship_name, ship_class, 200, 100, [laser, laser, laser])
        else:
            return "Unknown ship class"
        
        fleet.add_ship(ship)
        return f"Built {ship_name} and added to {fleet_name}"
    
    def simulate_combat(self, fleet1_name, fleet2_name):
        """Simulate combat between two fleets"""
        fleet1 = None
        fleet2 = None
        
        for fleet in self.empire.fleets:
            if fleet.name.lower() == fleet1_name.lower():
                fleet1 = fleet
            elif fleet.name.lower() == fleet2_name.lower():
                fleet2 = fleet
        
        if not fleet1 or not fleet2:
            return "One or both fleets not found"
        
        combat = Combat(fleet1, fleet2)
        winner = combat.resolve()
        
        return combat.get_log()
    
    def get_empire_status(self):
        """Get current empire status"""
        status = self.empire.get_status()
        
        # Add resource info
        status["resources"] = {
            "minerals": self.empire.resources.get(ResourceType.MINERALS),
            "energy": self.empire.resources.get(ResourceType.ENERGY),
            "population": self.empire.resources.get(ResourceType.POPULATION),
            "research_points": self.empire.resources.get(ResourceType.RESEARCH_POINTS),
        }
        
        return status
    
    def list_systems(self):
        """List all star systems"""
        return self.galaxy.systems
    
    def list_fleets(self):
        """List all fleets"""
        return self.empire.fleets

File: src/game.cpp
#include "game.h"
#include <algorithm>
#include <cctype>

Game::Game(const std::string& empireName)
    : empire(std::make_shared<Empire>(empireName)),
      galaxy(std::make_shared<Galaxy>(20)),
      running(false) {
    setupGame();
}

void Game::setupGame() {
    // Colonize home planet (Earth)
    auto homePlanets = galaxy->getHomeSystem()->getPlanets();
    if (homePlanets.size() >= 3) {
        auto homePlanet = homePlanets[2];  // 3rd planet
        auto earthColony = std::make_shared<Colony>("Earth", homePlanet);
        homePlanet->colonize(earthColony);
        empire->addColony(earthColony);
    }
    
    // Create starting fleet
    auto startingFleet = createStartingFleet();
    empire->addFleet(startingFleet);
}

std::shared_ptr<Fleet> Game::createStartingFleet() {
    auto fleet = std::make_shared<Fleet>("Home Defense Fleet", empire->getName());
    
    // Add basic ships
    Weapon laser("Laser Cannon", 10, 0.7, 5);
    
    auto scout = std::make_shared<Ship>("Scout-1", ShipClass::SCOUT, 50, 20,
                                       std::vector<Weapon>{laser});
    auto corvette = std::make_shared<Ship>("Corvette-1", ShipClass::CORVETTE, 100, 50,
                                          std::vector<Weapon>{laser, laser});
    
    fleet->addShip(scout);
    fleet->addShip(corvette);
    fleet->setLocation(galaxy->getHomeSystem());
    
    return fleet;
}

std::string Game::advanceTurn() {
    return empire->advanceTurn();
}

std::string Game::exploreSystem(const std::string& systemName) {
    const auto toLowerChar = [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    };

    for (auto& system : galaxy->getSystems()) {
        std::string sysNameLower = system->getName();
        std::string searchNameLower = systemName;
        std::transform(sysNameLower.begin(), sysNameLower.end(), sysNameLower.begin(), toLowerChar);
        std::transform(searchNameLower.begin(), searchNameLower.end(), searchNameLower.begin(), toLowerChar);
        
        if (sysNameLower == searchNameLower) {
            const bool wasExplored = system->isExplored();
            system->explore();

            if (!wasExplored) {
                const int reward = 10 + static_cast<int>(system->getPlanets().size()) * 2;
                empire->getResources().add(ResourceType::RESEARCH_POINTS, reward);
                return "Explored " + system->getName() + "! Found " +
                       std::to_string(system->getPlanets().size()) + " planets. Gained " +
                       std::to_string(reward) + " research points.";
            }

            return "System already explored: " + system->getName();
        }
    }
    return "System not found";
}

std::string Game::startResearch(const std::string& techId) {
    if (empire->setResearch(techId)) {
        auto tech = empire->getResearch().getTech(techId);
        if (tech) {
            return "Now researching: " + tech->getName();
        }
    }
    return "Cannot research that technology";
}

std::vector<std::shared_ptr<Technology>> Game::getAvailableResearch() {
    return empire->getResearch().getAvailableTechs();
}

std::string Game::buildShip(ShipClass shipClass, const std::string& fleetName) {
    // Find fleet
    std::shared_ptr<Fleet> targetFleet;
    std::string fleetNameLower = fleetName;
    const auto toLowerChar = [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    };
    std::transform(fleetNameLower.begin(), fleetNameLower.end(), fleetNameLower.begin(), toLowerChar);
    
    for (auto& fleet : empire->getFleets()) {
        std::string fname = fleet->getName();
        std::transform(fname.begin(), fname.end(), fname.begin(), toLowerChar);
        if (fname == fleetNameLower) {
            targetFleet = fleet;
            break;
        }
    }
    
    if (!targetFleet) {
        return "Fleet not found";
    }
    
    // Simple ship creation
    std::string shipName = shipClassToString(shipClass) + "-" + 
                          std::to_string(targetFleet->getShips().size() + 1);
    
    Weapon laser("Laser", 10, 0.7, 5);
    std::shared_ptr<Ship> ship;
    
    if (shipClass == ShipClass::SCOUT) {
        ship = std::make_shared<Ship>(shipName, shipClass, 50, 20,
                                     std::vector<Weapon>{laser});
    } else if (shipClass == ShipClass::CORVETTE) {
        ship = std::make_shared<Ship>(shipName, shipClass, 100, 50,
                                     std::vector<Weapon>{laser, laser});
    } else if (shipClass == ShipClass::FRIGATE) {
        Weapon betterLaser("Heavy Laser", 15, 0.75, 6);
        ship = std::make_shared<Ship>(shipName, shipClass, 200, 100,
                                     std::vector<Weapon>{betterLaser, betterLaser, betterLaser});
    } else {
        return "Unknown ship class";
    }
    
    targetFleet->addShip(ship);
    return "Built " + shipName + " and added to " + targetFleet->getName();
}

std::string Game::simulateCombat(const std::string& fleet1Name, const std::string& fleet2Name) {
    std::shared_ptr<Fleet> fleet1, fleet2;

    const auto toLowerChar = [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    };
    
    std::string f1Lower = fleet1Name, f2Lower = fleet2Name;
    std::transform(f1Lower.begin(), f1Lower.end(), f1Lower.begin(), toLowerChar);
    std::transform(f2Lower.begin(), f2Lower.end(), f2Lower.begin(), toLowerChar);
    
    for (auto& fleet : empire->getFleets()) {
        std::string fname = fleet->getName();
        std::transform(fname.begin(), fname.end(), fname.begin(), toLowerChar);
        if (fname == f1Lower) fleet1 = fleet;
        if (fname == f2Lower) fleet2 = fleet;
    }
    
    if (!fleet1 || !fleet2) {
        return "One or both fleets not found";
    }
    
    Combat combat(fleet1, fleet2);
    combat.resolve();
    
    std::string result;
    for (const auto& log : combat.getLog()) {
        result += log + "\n";
    }
    return result;
}
